pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,239 @@
+	errorIssuerInit        = "IssuerInitError"
+	errorSavingCertificate = "SaveCertError"
+
+	errorCertificateSigningRequestNotFound = "CSRNotFound"
+	errorCertificateSigningRequestParse    = "CSRParseError"
+
+	errorCertificateNotExists = "CertNotExists"
+	errorCertificateParse     = "CertParseError"
+
+	reasonIssuingCertificate = "IssueCert"
+	successCertificateIssued = "CertIssued"
+
+	messageErrorSavingCertificate = "Error saving TLS certificate: "
+)
+
+var (
+	certificateRequestGvk = v1alpha1.SchemeGroupVersion.WithKind("CertificateRequest")
+)
+
+func (c *controller) Sync(ctx context.Context, cr *v1alpha1.CertificateRequest) (err error) {
+	c.metrics.IncrementSyncCallCount(ControllerName)
+
+	log := logf.FromContext(ctx)
+	dbg := log.V(logf.DebugLevel)
+
+	crCopy := cr.DeepCopy()
+	defer func() {
+		if _, saveErr := c.updateCertificateRequestStatus(ctx, cr, crCopy); saveErr != nil {
+			err = utilerrors.NewAggregate([]error{saveErr, err})
+		}
+	}()
+
+	dbg.Info("Fetching existing certificate signing request and certificate from certificate request",
+		"name", crCopy.ObjectMeta.Name)
+	if len(cr.Spec.CSRPEM) == 0 {
+		return errors.New(errorCertificateSigningRequestNotFound)
+	}
+
+	csr, err := pki.DecodeX509CertificateRequestBytes(cr.Spec.CSRPEM)
+	if err != nil {
+		return err
+	}
+
+	el := validation.ValidateCertificateRequest(crCopy)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	// step zero: check if the referenced issuer exists and is ready
+	issuerObj, err := c.helper.GetGenericIssuer(crCopy.Spec.IssuerRef, crCopy.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	dbg.Info("Fetched issuer resource referenced by certificate request", "issuer_name", crCopy.Spec.IssuerRef.Name)
+
+	issuerReady := apiutil.IssuerHasCondition(issuerObj, v1alpha1.IssuerCondition{
+		Type:   v1alpha1.IssuerConditionReady,
+		Status: v1alpha1.ConditionTrue,
+	})
+	if !issuerReady {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotReady, "Issuer %s not ready", issuerObj.GetObjectMeta().Name)
+		return nil
+	}
+
+	i, err := c.issuerFactory.IssuerFor(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerInit, "Internal error initialising issuer: %v", err)
+		return nil
+	}
+
+	el = validation.ValidateCertificateRequestForIssuer(crCopy, issuerObj)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	if len(cr.Status.Certificate) == 0 {
+		dbg.Info("Invoking sign function as existing certificate does not exist")
+		return c.sign(ctx, i, crCopy, csr)
+	}
+
+	cert, err := pki.DecodeX509CertificateBytes(cr.Status.Certificate)
+	if err != nil {
+		return err
```

Comment:
Returning an error here without actually displaying anything to the user also seems like a dangerous move - we could endlessly call the `sign` function, and the user wouldn't know unless an administrator checks CMs logs.